scipy/special/_basic.py

```diff
@@ -2306,23 +2306,98 @@ def perm(N, k, exact=False):
 
 
 # https://stackoverflow.com/a/16327037
-def _range_prod(lo, hi):
+def _range_prod(lo, hi, k=1):
```

Comment:
On my machine this function from SO is consistently slower for anything less than 600 for exact computations. This is the code I am using, 

```python

import numpy as np
import perfplot


def f(n):
    '''Returns factorial of n'''
    r = 1
    for i in range(1, n + 1):
        r *= i
    return r


def range_prod(lo, hi):
    if lo+1 < hi:
        mid = (hi+lo)//2
        return range_prod(lo, mid) * range_prod(mid+1, hi)
    if lo == hi:
        return lo
    return lo*hi


def f2(n):
    if n < 2:
        return 1
    return range_prod(1, n)

perfplot.show(
    setup=lambda n: n,  # or setup=np.random.rand
    kernels=[f, f2],
    labels=["f", "f2"],
    n_range=[3, 5, 10, 20, 35, 50, 75, 100, 300, 500, 1000, 2000, 5000, 10000],
    xlabel="n",
    # logx="auto",  # set to True or False to force scaling
    # logy="auto",
    equality_check=None,  # set to None to disable "correctness" assertion
)

```

![image](https://user-images.githubusercontent.com/1303842/160273507-e1dc2cd5-0ab7-464e-8703-74bc96d64ae5.png)

if we zoom in to the origin 

![image](https://user-images.githubusercontent.com/1303842/160273516-3e97a6c9-efc0-4094-959f-da811ea8b36f.png)

So maybe it needs a switch if this is an important speed difference (for small n it is ~3X fast). I don't have any preference.

Reply:
Thanks for testing this so thoroughly!

Since the function is pre-existing (and I'm just teaching it the `k>1` case), I'm kind of reluctant to add one more thing to this PR, because it's already verging on the large side and finding the right cross-overpoint might take a bunch more analysis depending on platform and other issues. I think it would be better suited for a follow-up.